### DISCUSSION TITLE
Will Supabase fit my needs? (can I cram diverse objects into one column)

### DISCUSSION THREAD

**Discussion by noconsulate** [supabase/supabase#1079] [Questions] [answered]
I've been messing around with a few different backend/database stacks and found that Parse was fitting my needs. But now I have discovered Supbase and it looks very promising and more active. Before I switch gears and try it out in my project I'd like to get one question about my use case answered so that I don't end up wasting time. 

Can I have a column that contains a deeply nested object where each instance will have different dimensions/properties?

I'm making a golf scoring app for multiple players and each game is a row. The 'scores' column is an object that varies depending on the number of players and number of holes. For instance 3 hole, 2 player game is like this:
`scores: {
1: {
1: null,
2: null
},
2: {
1: null,
2: null
},
3: {
1: null,
2: null,
}
}
`

and a 2 hole, 4 player game looks like:
`scores: {
1: {
1: null,
2: null,
3: null,
4: null
},
2: {
1: null,
2: null,
3: null,
4: null
}
}`

Now this is very easy to do in a document database but SQL... not so much. I'm sure it's quite possible with a bit of cleverness but I am completely ignorant about how relational databases work so anything I come up with will probably be poor solution, not to mention it'll take me for ever and drive me nuts. I was able to do my naive approach to my data very easily with Firebase, so since Supabase is meant to be a Firebase replacement maybe it'll work? Does Supabase have abstractions that will let me treat SQL like a document database?

**Comment by atnegrete** [comment, marked as answer]
I can certainly help you translate your NoSQL data structure to relational, if you explain a bit more what the fields are and how they are related. I prefer relational (almost)  any time over document.

At first glance, I'm guessing you'd have:

**matches** table
- id : primary key auto gen

**scores** table
- id: primary key auto gen
- matchId: FK to **matches**
- userId: FK to **users**
- score: int ( the actual score for this user, in this match)

Now, to answer your question, **yes** you can store a json type column. Just setup your keys/ids in a way you can still retrieve your data easily to how your application needs it.

**Comment by noconsulate** [comment]
Thanks for the answer, both of you. 

The relations for the data should be pretty easy to figure out. Perhaps it is a good idea to put give **scores** its own table and relate it to **matches**. **Scores** is still a nested object because there are 4 * 2 different configurations of scorecards. (Each combination of of players and holes yields a different matrix of scores, just like on a physical scorecard.) 

The **scores** are updated one player/hole at a time and the updates are sent to all the players in the game in real time. So another consideration is will I have to queary/update the entire **scores** column/field each time a score is updated or can I target just one field. 

This is actually easy mode and the scoring data will become more complicated. One player is keeping score for everyone in the game. Eventually I'll make a version where players are paired up and keep track of scores for themselves and their partner and differences will be highlighted in order to be fixed.

Given all that, do you see a way to represent this data without resorting to a nested object?

**Reply by atnegrete** [reply]
Sorry, I'm actually still not completely following how it's all laid out (and I don't play golf 😢 ).

It sounds like you're on the right track.

You could add a "hole" column to **scores** and then you could pre-populate them when the game starts with all players have all the different scorecards created (scores), and you could set **scores.score** to nullable so that you know if it's been set or not?

Hope that helps! Sorry, again, I don't play golf

**Reply by noconsulate** [reply]
Hey, I don't play golf either! But my friend/business partner does. 

Are you suggesting that each match has it's own **scores** table? It's easy to conceive of the scores for an individual match as a table, in fact here is what a scorecard looks like:
[scorecard](https://www.golfcrosswinds.com/img/scorecard-1.png)

Taking this example you have holes as columns and players as rows. If it's a 9-hole game there will be 9 columns and 18-hole game has 18 columns. So can I represent a scorecard as a table and link to a **scorecard** table in an individual cell in the **matches** table? This seems wrong to me due to my understanding that a table is a collection of data and I can't have potentially thousands of different tables, one table for each match. 

Or perhaps you're suggesting that one **scores** has columns for individual holes, and the rows would be identified by "**matches**.id + player"? Or maybe it's better to have one column to identify the match and another column to identify the player.  Or I could skip giving each player their own row and instead put an array of scores in each "hole" column. This makes sense!

Thanks for your help so far... using tables is suddenly a lot less scary. 

PS - My main conceptual problem is I don't understand how to represent multiple dimensions of data in SQL. For an example in keeping with golf, I'm hopefully going to be buying a big collection of golf course data for my app. I imagine it as a table with each golf course as a row. Part of the data for each course is a scorecard, exactly the data presented as a real-world table in the image linked above. My mind keeps wanting to see each scorecard as its own table but I don't think this is possibly right. I can also imagine that each cell in the scorecard has its own row in the Big overall table but that doesn't seem right either. I clearly don't get how abstractions work in SQL. I hope this makes sense, but I don't expect anyone to make me "get" SQL in a forum post. This is why a document database is so attractive... it's effortless to map my data as a nested object.

**Reply by atnegrete** [reply]
Oh, no certainly having a to create new tables to represent the same data again is never the right option. I think we're a bit on a different page, based on that scorecard image here's how I'd probably set it up. 

**matches** table
id : primary key auto gen

**scores** table
id: primary key auto gen
matchId: FK to matches
userId: FK to users
holeId: FK to holes
score: int (nullable - the actual score for this user, in this match, for this hole)
elapsedTime: int

**holes** table
id: primary key auto gen
name: string

When a match/game/round, whatever it's called, starts you can essentially pre-populate the **scores** table.  First you can create the match, and then insert the **holes** (which are basically just names?). Then, for each player insert as many rows as you need for the scorecard, so if it's 5 holes and 9 possible scores per hole, then 5 x 9 = 45 score rows needed. Set the same matchId for both players for every score. You can set the score value to null, and update it as the scores come in. Does that help?

Not always, but generally speaking, if you are thinking about document based.. You can typically assume whenever you have a nested array, that pretty much is another table and you have a foreign key to the parent document (aka table) to create the one to many relationship.

**Reply by noconsulate** [reply]
Yes, this is helpful. Thank you. I may not do it exactly as you suggested but I have a pretty good idea of how to think about this.

**Comment by burggraf** [comment]
You should look into using a JSONB column.  JSONB essentially gives you all the flexibility of a document database (JSON) in a single Postgresql column.  You can literally just store the examples you posted inside the database.  You can also index JSONB content and query it.
